File: utils/rendering/cache.py
import asyncio
import hashlib
import json
from pathlib import Path
import re
import shutil
import time
from weakref import WeakValueDictionary
from utils.paths import shared_cache
from utils.json_io import atomic_write_bytes, atomic_write_json
from .models import RenderedDocument, RenderedPage

_locks = WeakValueDictionary()


async def cached_render(key: str, render, *, force=False):
    lock = _locks.setdefault(key, asyncio.Lock())
    async with lock:
        directory = shared_cache("rendering") / key
        manifest = directory / "pages.json"
        if not force:
            try:
                items = json.loads(manifest.read_text())
                pages = []
                for i, item in enumerate(items):
                    data = (directory / f"{i}.png").read_bytes()
                    if hashlib.sha256(data).hexdigest() != item["sha256"]:
                        raise ValueError("Invalid cached page")
                    pages.append(RenderedPage(data, item["width"], item["height"]))
                if pages:
                    manifest.touch()
                    return RenderedDocument(tuple(pages))
            except (OSError, ValueError, KeyError, TypeError):
                pass
        result = await render()
        if not result.pages:
            raise ValueError("Renderer produced no pages")
        # Publish the manifest last: readers never accept a partial document.
        manifest.unlink(missing_ok=True)
        records = []
        for i, page in enumerate(result.pages):
            await asyncio.to_thread(
                atomic_write_bytes, directory / f"{i}.png", page.data
            )
            records.append(
                {
                    "width": page.width,
                    "height": page.height,
                    "sha256": hashlib.sha256(page.data).hexdigest(),
                }
            )
        await asyncio.to_thread(atomic_write_json, manifest, records)
        return result


def cleanup_render_cache(now=None):
    root = shared_cache("rendering")
    if not root.exists():
        return
    cutoff = (time.time() if now is None else now) - 7 * 86400
    for directory in root.iterdir():
        if not re.fullmatch(r"[0-9a-f]{64}", directory.name) or directory.is_symlink():
            continue
        lock = _locks.get(directory.name)
        if lock and lock.locked():
            continue
        manifest = directory / "pages.json"
        if manifest.is_file() and manifest.stat().st_mtime < cutoff:
            shutil.rmtree(directory)
            _locks.pop(directory.name, None)

File: utils/rendering/cache.go
package rendering

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sync"
	"time"

	"github.com/HakuchumuHYX/HakuBot/utils"
)

type Cache struct {
	Directory string
	mu        sync.Mutex
}
type cachedPage struct {
	Width, Height int
	SHA256        string
}

var cacheKey = regexp.MustCompile(`^[0-9a-f]{64}$`)

func (c *Cache) Render(ctx context.Context, key string, force bool, render func() (Document, error)) (Document, error) {
	if !cacheKey.MatchString(key) {
		return Document{}, errors.New("invalid render cache key")
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := ctx.Err(); err != nil {
		return Document{}, err
	}
	dir := filepath.Join(c.Directory, key)
	manifest := filepath.Join(dir, "pages.json")
	if !force {
		var entries []cachedPage
		if utils.ReadJSON(manifest, &entries) == nil && len(entries) > 0 {
			var d Document
			valid := true
			for i, e := range entries {
				data, err := os.ReadFile(filepath.Join(dir, fmt.Sprintf("%d.png", i)))
				sum := sha256.Sum256(data)
				if err != nil || hex.EncodeToString(sum[:]) != e.SHA256 {
					valid = false
					break
				}
				d.Pages = append(d.Pages, Page{data, e.Width, e.Height})
			}
			if valid {
				now := time.Now()
				if err := os.Chtimes(manifest, now, now); err != nil {
					return Document{}, err
				}
				return d, nil
			}
		}
	}
	d, err := render()
	if err != nil {
		return d, err
	}
	if len(d.Pages) == 0 {
		return d, errors.New("renderer returned no pages")
	}
	if err = os.Remove(manifest); err != nil && !errors.Is(err, os.ErrNotExist) {
		return Document{}, err
	}
	var entries []cachedPage
	for i, p := range d.Pages {
		if err = utils.AtomicWrite(filepath.Join(dir, fmt.Sprintf("%d.png", i)), p.Data); err != nil {
			return Document{}, err
		}
		sum := sha256.Sum256(p.Data)
		entries = append(entries, cachedPage{p.Width, p.Height, hex.EncodeToString(sum[:])})
	}
	return d, utils.WriteJSON(manifest, entries)
}
func (c *Cache) Cleanup(now time.Time) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	entries, err := os.ReadDir(c.Directory)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	if err != nil {
		return err
	}
	for _, e := range entries {
		if !e.IsDir() || !cacheKey.MatchString(e.Name()) {
			continue
		}
		dir := filepath.Join(c.Directory, e.Name())
		info, err := os.Stat(filepath.Join(dir, "pages.json"))
		if errors.Is(err, os.ErrNotExist) {
			continue
		}
		if err != nil {
			return err
		}
		if now.Sub(info.ModTime()) > 7*24*time.Hour {
			if err = os.RemoveAll(dir); err != nil {
				return err
			}
		}
	}
	return nil
}
